nextest-filtering/src/lib.rs

```diff
@@ -0,0 +1,271 @@
+    match matcher {
+        RawNameMatcher::Contains(t) => Some(NameMatcher::Contains(t.clone())),
+        RawNameMatcher::Equal(t) => Some(NameMatcher::Equal(t.clone())),
+        RawNameMatcher::Regex(r) => Some(NameMatcher::Regex(r.clone())),
+        RawNameMatcher::Error => None,
+    }
+}
+
+fn compile_set_def(
+    set: &SetDef,
+    packages: &[PackageMetadata],
+    cache: &mut DependsCache,
+) -> Option<FilteringSet> {
+    match set {
+        SetDef::Package(matcher) => Some(FilteringSet::Packages(matching_packages(
+            &to_name_matcher(matcher)?,
+            packages,
+        ))),
+        SetDef::Deps(matcher) => Some(FilteringSet::Packages(dependencies_packages(
+            &to_name_matcher(matcher)?,
+            packages,
+            cache,
+        ))),
+        SetDef::Rdeps(matcher) => Some(FilteringSet::Packages(rdependencies_packages(
+            &to_name_matcher(matcher)?,
+            packages,
+            cache,
+        ))),
+        SetDef::Test(matcher) => Some(FilteringSet::Test(to_name_matcher(matcher)?)),
+        SetDef::All => Some(FilteringSet::All),
+        SetDef::None => Some(FilteringSet::None),
+        SetDef::Error => None,
+    }
+}
+
+fn compile_expr(
+    expr: &Expr,
+    packages: &[PackageMetadata],
+    cache: &mut DependsCache,
+) -> Option<FilteringExpr> {
+    match expr {
+        Expr::Set(set) => Some(FilteringExpr::Set(compile_set_def(set, packages, cache)?)),
+        Expr::Not(expr) => Some(FilteringExpr::Not(Box::new(compile_expr(
+            expr, packages, cache,
+        )?))),
+        Expr::Union(expr_1, expr_2) => Some(FilteringExpr::Union(
+            Box::new(compile_expr(expr_1, packages, cache)?),
+            Box::new(compile_expr(expr_2, packages, cache)?),
+        )),
+        Expr::Intersection(expr_1, expr_2) => Some(FilteringExpr::Intersection(
+            Box::new(compile_expr(expr_1, packages, cache)?),
+            Box::new(compile_expr(expr_2, packages, cache)?),
+        )),
+        Expr::Error => None,
+    }
+}
+
+impl FilteringExpr {
+    /// Parse a filtering expression
+    pub fn parse(
+        input: &str,
+        graph: &PackageGraph,
+    ) -> Result<FilteringExpr, error::FilteringExprParsingError> {
+        let errors = RefCell::new(Vec::new());
+        match parsing::parse(parsing::Span::new_extra(input, error::State::new(&errors))) {
+            Ok(expr) => {
+                let errors = errors.into_inner();
+
+                if !errors.is_empty() {
+                    for err in errors {
+                        let report = miette::Report::new(err).with_source_code(input.to_string());
+                        eprintln!("{:?}", report);
+                    }
+                    return Err(error::FilteringExprParsingError(input.to_string()));
+                }
+
+                let in_workspace_packages: Vec<_> =
+                    graph.packages().filter(|p| p.in_workspace()).collect();
```

Comment:
You can just do `graph.resolve_workspace().packages(DependencyDirection::Forward).collect()` which will avoid an iteration over the full package graph. Not a huge deal.